### PR TITLE
plan: rewrite exact `like` expression to equal condition

### LINKED ISSUE #8962: the optimizer should treat predicate 'like' the same way as predicate '=' when it is an exact match
there is a multi-column index (a,c,b) on the table, when using 'like' on a condition which is an exact match, the next column in multi-column index could not use index scan.

```
create table t3 (a int, b decimal(10,2), c varchar(10), d double, e timestamp);
desc select * from t3 where a=5 and b=5.2 and c like 'ok';
desc select * from t3 where a=5 and b=5.2 and c = 'ok';
```
```
mysql> desc select * from t3 where a=5 and b=5.2 and c like 'ok';
+-----------------------+-------+------+--------------------------------------------------------------------------------+
| id                    | count | task | operator info                                                                  |
+-----------------------+-------+------+--------------------------------------------------------------------------------+
| IndexLookUp_11        | 0.10  | root |                                                                                |
| ├─Selection_10        | 0.10  | cop  | eq(gin.t3.b, 5.2)                                                              |
| │ └─IndexScan_8       | 0.10  | cop  | table:t3, index:a, c, b, range:[5 "ok",5 "ok"], keep order:false, stats:pseudo |
| └─TableScan_9         | 0.10  | cop  | table:t3, keep order:false                                                     |
+-----------------------+-------+------+--------------------------------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> desc select * from t3 where a=5 and b=5.2 and c = 'ok';
+-------------------+-------+------+------------------------------------------------------------------------------------------+
| id                | count | task | operator info                                                                            |
+-------------------+-------+------+------------------------------------------------------------------------------------------+
| IndexLookUp_10    | 0.00  | root |                                                                                          |
| ├─IndexScan_8     | 0.00  | cop  | table:t3, index:a, c, b, range:[5 "ok" 5.20,5 "ok" 5.20], keep order:false, stats:pseudo |
| └─TableScan_9     | 0.00  | cop  | table:t3, keep order:false, stats:pseudo                                                 |
+-------------------+-------+------+------------------------------------------------------------------------------------------+
3 rows in set (0.00 sec)
```

tidb version:

> Release Version: v2.1.2-1-g8ba8096
> Git Commit Hash: 8ba809622668a1287e9f7152bfd34ad6be220304
> Git Branch: release-2.1
> UTC Build Time: 2018-12-21 03:45:55
> GoVersion: go version go1.11.2 linux/amd64
> Race Enabled: false
> TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
> Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by hhxcc**
### What problem does this PR solve? 
 Fix #8962 

### What is changed and how it works?
rewrite such kind of like function to equal function.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by winoros** [comment]
Thanks for your contribution!

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -31,6 +28,9 @@ import (
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/parser_driver"
 	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/stringutil"
+	"strconv"
+	"strings"
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]: Please keep the builtin packages in the front as other files do.

File: util/stringutil/string_util_test.go [hunk 2]
```diff
@@ -124,3 +124,22 @@ func (s *testStringUtilSuite) TestPatternMatch(c *C) {
...
+		exactMatch bool
+	}{
+		{nil, true},
+		{[]byte{}, true},
+		{[]byte{1}, true},
+		{[]byte{1, 1}, true},
+		{[]byte{1, 2}, false},
+		{[]byte{1, 3}, false},
+		{[]byte{1, 2, 3}, false},
+	}
```
Comment (eurekaka) [inline util/stringutil/string_util_test.go]: Better to use original string as input and `CompilePattern` to convert the string, then use the result to test `IsExactMatch`.

File: planner/core/expression_rewriter.go [hunk 3]
```diff
@@ -31,6 +28,9 @@ import (
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/parser_driver"
 	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/stringutil"
+	"strconv"
```
[the remark this answers is not in the document]
Comment (hhxcc) [reply planner/core/expression_rewriter.go]: ok

File: planner/core/expression_rewriter.go [hunk 4]
```diff
@@ -1129,16 +1129,39 @@ func (er *expressionRewriter) caseToExpression(v *ast.CaseExpr) {
...
+
+	var fieldType = &types.FieldType{}
+	var function expression.Expression
+	// treat predicate 'like' the same way as predicate '=' when it is an exact match.
+	var isPatternExactMatch = false
+	if pat, ok := v.Pattern.(*driver.ValueExpr); ok {
+		if patValue, ok := pat.GetValue().(string); ok {
+			patValue, patTypes := stringutil.CompilePattern(patValue, v.Escape)
+			if stringutil.IsExactMatch(patTypes) {
+				op := opcode.EQ
```
[the remark this answers is not in the document]
Comment (hhxcc) [reply planner/core/expression_rewriter.go]: ok

**Comment by hhxcc** [comment]
@winoros  when pattern string contains escape character , we should use the compiled result of the pattern string  in the equal condition,  eg:   `a LIKE 'ab\_c'  `  should rewrite to  `a =  'ab_c'  `. 
So,  we can not  just check whether er.ctxStack[l-1] is a string without regex pattern.

File: util/stringutil/string_util_test.go (hunk 2 above)
Comment (hhxcc) [reply util/stringutil/string_util_test.go]: ok, Done.

File: planner/core/expression_rewriter.go [hunk 5]
```diff
@@ -1129,16 +1129,39 @@ func (er *expressionRewriter) caseToExpression(v *ast.CaseExpr) {
...
-	types.DefaultTypeForValue(int(v.Escape), escapeTp)
-	function := er.notToExpression(v.Not, ast.Like, &v.Type,
-		er.ctxStack[l-2], er.ctxStack[l-1], &expression.Constant{Value: types.NewIntDatum(int64(v.Escape)), RetType: escapeTp})
+
+	var fieldType = &types.FieldType{}
+	var function expression.Expression
+	// treat predicate 'like' the same way as predicate '=' when it is an exact match.
+	var isPatternExactMatch = false
+	if pat, ok := v.Pattern.(*driver.ValueExpr); ok {
+		if patValue, ok := pat.GetValue().(string); ok {
```
[the remark this answers is not in the document]
Comment (eurekaka) [reply planner/core/expression_rewriter.go]: @hhxcc please address this comment, `expressionRewriter` should have traversed the parameters of `like` recursively already, so `v.Pattern` should have been rewritten already and saved on the stack, then you can directly pop it out from the stack and use it.

File: planner/core/expression_rewriter.go [hunk 6]
```diff
@@ -1129,16 +1130,39 @@ func (er *expressionRewriter) caseToExpression(v *ast.CaseExpr) {
...
 		return
 	}
-	escapeTp := &types.FieldType{}
-	types.DefaultTypeForValue(int(v.Escape), escapeTp)
-	function := er.notToExpression(v.Not, ast.Like, &v.Type,
-		er.ctxStack[l-2], er.ctxStack[l-1], &expression.Constant{Value: types.NewIntDatum(int64(v.Escape)), RetType: escapeTp})
+
+	var fieldType = &types.FieldType{}
+	var function expression.Expression
+	// treat predicate 'like' the same way as predicate '=' when it is an exact match.
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]:
```suggestion
	// Treat predicate 'like' the same way as predicate '=' when it is an exact match.
```

File: planner/core/expression_rewriter.go [hunk 7]
```diff
@@ -1129,16 +1130,39 @@ func (er *expressionRewriter) caseToExpression(v *ast.CaseExpr) {
...
 	er.err = expression.CheckArgsNotMultiColumnRow(er.ctxStack[l-2:]...)
 	if er.err != nil {
 		return
 	}
-	escapeTp := &types.FieldType{}
-	types.DefaultTypeForValue(int(v.Escape), escapeTp)
-	function := er.notToExpression(v.Not, ast.Like, &v.Type,
-		er.ctxStack[l-2], er.ctxStack[l-1], &expression.Constant{Value: types.NewIntDatum(int64(v.Escape)), RetType: escapeTp})
+
+	var fieldType = &types.FieldType{}
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]:
```suggestion
	fieldType := &types.FieldType{}
```

File: planner/core/expression_rewriter.go [hunk 8]
```diff
@@ -1129,16 +1130,39 @@ func (er *expressionRewriter) caseToExpression(v *ast.CaseExpr) {
...
+	var function expression.Expression
+	// treat predicate 'like' the same way as predicate '=' when it is an exact match.
+	var isPatternExactMatch = false
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]:
```suggestion
	isPatternExactMatch := false
```

File: planner/core/expression_rewriter.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hhxcc) [reply planner/core/expression_rewriter.go]: ok, done.

File: planner/core/expression_rewriter.go [hunk 9]
```diff
@@ -1129,16 +1130,39 @@ func (er *expressionRewriter) caseToExpression(v *ast.CaseExpr) {
...
-	types.DefaultTypeForValue(int(v.Escape), escapeTp)
-	function := er.notToExpression(v.Not, ast.Like, &v.Type,
-		er.ctxStack[l-2], er.ctxStack[l-1], &expression.Constant{Value: types.NewIntDatum(int64(v.Escape)), RetType: escapeTp})
+
+	var function expression.Expression
+	fieldType := &types.FieldType{}
+	isPatternExactMatch := false
+	// Treat predicate 'like' the same way as predicate '=' when it is an exact match.
+	if patExpression, ok := er.ctxStack[l-1].(*expression.Constant); ok {
+		if patExpression.RetType.EvalType() == types.ETString {
```
Comment (eurekaka) [inline planner/core/expression_rewriter.go]: Should we use `patExpression.EvalString()` to get `string`?

Comment (hhxcc) [reply planner/core/expression_rewriter.go]: ok

**Comment by winoros** [comment]
We need to add `ast.Like` to `expression.UnCacheableFunctions`?

**Comment by eurekaka** [comment]
> We need to add `ast.Like` to `expression.UnCacheableFunctions`?

Any specific reason?

**Pull request merged**